devolutions-crypto-wayk/src/bastion/fle.rs

```diff
@@ -0,0 +1,203 @@
+use crate::bastion::Error;
+use block_padding::{Padding, Pkcs7};
+use chacha20poly1305::aead::{AeadInPlace, NewAead};
+use chacha20poly1305::{Key, XChaCha20Poly1305, XNonce};
+use uuid::Uuid;
+
+pub const AAD_MAGIC_NUMBER: [u8; 4] = [0x53, 0x0A, 0x0F, 0x0E];
+pub const SAFE_PAYLOAD_PREFIX: &str = "|~Safe~|";
+pub const DEFAULT_BLOCK_SIZE: usize = 32;
+
+/// Encrypts field-level content using default padding size.
+///
+/// See `encrypt_field_with_block_size` for details.
+pub fn encrypt_field(master_key_id: Uuid, key: &[u8], content: &[u8]) -> Result<String, Error> {
+    encrypt_field_with_block_size(master_key_id, key, content, DEFAULT_BLOCK_SIZE)
+}
+
+/// Encrypts field-level content using provided padding size.
+///
+/// - Output is the encrypted content in Bastion safe Field Value Format (ie: |~Safe~|k1AjQZYciYvlPmJMnHjbb…)
+///
+/// # Interals
+///
+/// - AAD is composed of a 4-byte magic number, a 4-byte reserved field and the master key UUID in big-endian binary format.
+/// - A 24-byte initialization vector (IV) is randomly generated.
+/// - Content is padded using PKCS#7 padding mode.
+/// - Encryption using XChaCha2020-Poly1035.
+/// - Output is a buffer such as `[magic number (4) | reserved (4) | UUID (16) | IV (24) | Ciphertext (variable) | Tag (16)]`
+///     encoded in Bastion safe Field Value Format.
+pub fn encrypt_field_with_block_size(
+    master_key_id: Uuid,
+    key: &[u8],
+    content: &[u8],
+    block_size: usize,
+) -> Result<String, Error> {
+    use rand::rngs::OsRng;
+    use rand::Fill;
+
+    let pad_leftover = block_size - (content.len() % block_size);
+    let total_len = AAD_SIZE + IV_SIZE + content.len() + pad_leftover + TAG_SIZE;
+    let mut buffer = vec![0u8; total_len];
+
+    // associated data [magic number (4) | reserved (4) | UUID (16)]
+    let (aad, rest) = buffer.split_at_mut(AAD_SIZE);
+    {
+        let (aad_magic, aad_rest) = aad.split_at_mut(AAD_MAGIC_NUMBER_SIZE);
+        aad_magic.copy_from_slice(&AAD_MAGIC_NUMBER);
+        let (aad_reserved, aad_uuid) = aad_rest.split_at_mut(AAD_RESERVED_SIZE);
+        aad_reserved.copy_from_slice(&[0, 0, 0, 0]); // useless, but just to be extra safe…
+        aad_uuid.copy_from_slice(master_key_id.as_bytes());
+    }
+
+    // nonce / IV
+    let (nonce, rest) = rest.split_at_mut(IV_SIZE);
+    nonce.try_fill(&mut OsRng)?;
+    let nonce = XNonce::from_slice(nonce);
+
+    // actual data
+    let (data, rest) = rest.split_at_mut(rest.len() - TAG_SIZE);
+    data[..content.len()].copy_from_slice(&content);
+    Pkcs7::pad(data, content.len(), block_size).map_err(|_| Error::Padding)?;
```

Comment:
The result is not used, so the message is not padded here. (the operation is not in-place according to the doc, might be UB, need to investigate further)

Reply:
Operation is in-place. `Padding::pad` returns a mutable view on the input buffer modified in-place (useful when providing a buffer bigger than required to avoid calculating padded length twice). In fact `Padding::pad` is a helper around `Padding::pad_block` that I could use instead if you prefer: https://docs.rs/block-padding/0.2.1/src/block_padding/lib.rs.html#34

Reply:
I changed to directly use `Padding::pad_block` so there is no confusion